mleap-xgboost-runtime/src/test/scala/ml/combust/mleap/xgboost/runtime/XGBoostPredictorClassificationModelParitySpec.scala

```diff
@@ -38,7 +38,8 @@ class XGBoostPredictorClassificationModelParitySpec extends FunSpec
         assert(
           almostEqualSequences(
             Seq(boosterProbability.values),
-            Seq(mleapResult.dataset.head.getTensor[Double](mleapProbabilityColIndex).toArray)
+            Seq(mleapResult.dataset.head.getTensor[Double](mleapProbabilityColIndex).toArray),
+            precision = 1e-6
```

Comment:
needing to specify the precision level makes me nervous.  iirc, xgboost is always 32-bit floats, so using double's here might be the issue?  Or the predictor library also needs to be upgraded for the latest xgboost version?

Reply:
I see, this is helpful. Let me see if I can get rid of this.

Reply:
I dug further into this, and I'm hoping you can help me make sense of what I'm seeing. First, trying to use floats instead of doubles here is rather involved since the reason Doubles are used in the first place is so the code plays well with data structures such as Spark ML Vectors, for which there is no Float-specific implementation (only Double).

The test that fails that I'm trying to fix is [this one](https://github.com/combust/mleap/blob/61c15aa128fe06a2096b09007ad6f047c71fbb6f/mleap-xgboost-runtime/src/test/scala/ml/combust/mleap/xgboost/runtime/XGBoostPredictorClassificationModelParitySpec.scala#L117). I've noticed by fooling around that if I load the transformer using `loadMleapTransformerFromBundle` instead of `loadXGBoostPredictorFromBundle`, the error goes away and the predictions match exactly. Can you help me understand the difference between those two?

Reply:
The difference between `loadMleapTransformerFromBundle` and `loadXGBoostPredictorFromBundle` is which library is used in the backend.  There are two libraries for making predictions for xgboost models:

- The dmlc xgboost library.  This is the official open source code for xgboost, and its java/scala API is just a JNI wrapper around xgboost's main c++ code.  The JNI process introduces extra overhead and makes predictions for single rows be slow.
- The xgboost-predictor library is a java implementation that only allows for making predictions (i.e., you can't train an xgboost model with this library).  This library has been benchmarked as being several orders of magnitude lower latency for making predictions on single rows.

mleap has a special op so that folks can decide to use the predictor library instead of the dmlc xgboost.  Your observation that the predictions match exactly when using the loadMleapTransformerFromBundle (i.e., dmlc xgboost) but don't match when using loadXGBoostPredictorFromBundle (i.e., xgboost-predictor library) indicates that there is a problem which we need to fix.